src/theory/strings/infer_proof_cons.cpp

```diff
@@ -0,0 +1,1043 @@
+    case Inference::SSPLIT_CST:
+    case Inference::SSPLIT_VAR:
+    {
+      Trace("strings-ipc-core") << "Generate core rule for " << infer
+                                << " (rev=" << isRev << ")" << std::endl;
+      // All of the above inferences have the form:
+      //   <explanation for why t and s have the same prefix/suffix> ^
+      //   t = s ^
+      //  <length constraint>?
+      // We call t=s the "main equality" below. The length constraint is
+      // optional, which we split on below.
+      size_t nchild = ps.d_children.size();
+      size_t mainEqIndex = 0;
+      bool mainEqIndexSet = false;
+      // the length constraint
+      std::vector<Node> lenConstraint;
+      // these inferences have a length constraint as the last explain
+      if (infer == Inference::N_UNIFY || infer == Inference::F_UNIFY
+          || infer == Inference::SSPLIT_CST || infer == Inference::SSPLIT_VAR
+          || infer == Inference::SSPLIT_VAR_PROP
+          || infer == Inference::SSPLIT_CST_PROP)
+      {
+        if (exp.size() >= 2)
+        {
+          std::map<size_t, size_t>::iterator itsei =
+              startExpIndex.find(exp.size() - 1);
+          if (itsei != startExpIndex.end())
+          {
+            // The index of the "main" equality is the last equality before
+            // the length explanation.
+            mainEqIndex = itsei->second - 1;
+            mainEqIndexSet = true;
+            // the remainder is the length constraint
+            lenConstraint.insert(lenConstraint.end(),
+                                 ps.d_children.begin() + mainEqIndex + 1,
+                                 ps.d_children.end());
+          }
+        }
+      }
+      else
+      {
+        if (nchild >= 1)
+        {
+          // The index of the main equality is the last child.
+          mainEqIndex = nchild - 1;
+          mainEqIndexSet = true;
+        }
+      }
+      Node mainEq;
+      if (mainEqIndexSet)
+      {
+        mainEq = ps.d_children[mainEqIndex];
+        Trace("strings-ipc-core") << "Main equality " << mainEq << " at index "
+                                  << mainEqIndex << std::endl;
+      }
+      if (mainEq.isNull() || mainEq.getKind() != EQUAL)
+      {
+        Trace("strings-ipc-core")
+            << "...failed to find main equality" << std::endl;
+        break;
+      }
+      // apply MACRO_SR_PRED_ELIM using equalities up to the main eq
+      std::vector<Node> childrenSRew;
+      childrenSRew.push_back(mainEq);
+      childrenSRew.insert(childrenSRew.end(),
+                          ps.d_children.begin(),
+                          ps.d_children.begin() + mainEqIndex);
+      Node mainEqSRew =
+          psb.tryStep(PfRule::MACRO_SR_PRED_ELIM, childrenSRew, {});
+      if (CDProof::isSame(mainEqSRew, mainEq))
+      {
+        Trace("strings-ipc-core") << "...undo step" << std::endl;
+        // the rule added above was not necessary
+        psb.popStep();
```

Comment:
Couldn't that just be automatically done in `TheoryProofStepBuffer`?

Reply:
TheoryProofStepBuffer is not necessarily tied to CDProof.  Essentially, this code is here because the strings proof constructor is using a CDProof, which handles some inference steps automatically (those that return true for CDProof::isSame), but other uses of TheoryProofStepBuffer could be different.